### PR TITLE
[DPE-3358] Switch to `kafka` relation

### PR DESCRIPTION
**Pull request by zmraul**
Removes all config options and switches to use `kafka_client` interface.

NOTE: A bugfix is added to the `data_interfaces` lib. This should be removed and addressed by merging and updating the lib:
- https://github.com/canonical/data-platform-libs/pull/163
- https://github.com/canonical/data-platform-libs/pull/162

### CODE REVIEW

File: src/core/models.py [hunk 1]
```diff
@@ -270,3 +267,18 @@ def kafka_ready(self) -> bool:
             return False
 
         return True
+
+    def brokers_active(self) -> bool:
+        """Check that Kafka is active."""
+        # FIXME SSL connection info through models?
```
Comment (deusebio) [inline src/core/models.py:273]: **question** Can we elaborate more on this? can we maybe have a Jira ticket for this?

File: src/core/models.py [hunk 2]
```diff
@@ -75,13 +66,23 @@ def update(self, items: dict[str, str]) -> None:
         for field in delete_fields:
             del self.relation_data[field]
 
+    # def update(self, items: dict[str, str]) -> None:
```
Comment (deusebio) [inline src/core/models.py:69]: **todo** shall we remove this if not needed?

File: src/core/models.py [hunk 3]
```diff
@@ -35,30 +36,14 @@ def __init__(
...
-        relation: Relation,
-        data_interface: Data,
-        component: Unit | Application,
-        substrate: Substrate,
-    ):
-        super().__init__(relation, data_interface, component, substrate)
-        # Redundant definition as lint can't resolve that super's relation may be None
-        self.relation = relation
-        self.relation_data = self.data_interface.as_dict(self.relation.id)
+    def __bool__(self):
```
Comment (deusebio) [inline src/core/models.py:41]: **praise** I like this!

**Review by deusebio — APPROVED** [review]
Amazing! I really liked the PR! The changes are very well confined and structured. Well done!

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 4]
```diff
@@ -3051,7 +3051,7 @@ def _on_relation_created_event(self, event: RelationCreatedEvent) -> None:
 
         # Sets topic, extra user roles, and "consumer-group-prefix" in the relation
         relation_data = {
-            f: getattr(self, f.replace("-", "_"), "")
+            f: getattr(self.relation_data, f.replace("-", "_"), "")
```
Comment (marcoppenheimer) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:3054]: **question:** I recall you mentioning this in a meeting, what was this for again?

File: src/charm.py [hunk 5]
```diff
@@ -104,22 +95,26 @@ def _on_config_changed(self, event: ops.ConfigChangedEvent):
...
-
+            # Config is different, apply changes to file
             self.config_manager.generate_config()
 
-        if not self.workload.active():
-            self.workload.start()
-        else:
-            self.workload.restart()
+            # Restart so changes take effect
+            logger.info("Acquiring lock from _on_config_changed...")
```
Comment (marcoppenheimer) [inline src/charm.py:102]: **nitpick:** I think this log can be removed. I have removed it in Kafka now too, I think it was leftover debugging from ages ago.

**Review by marcoppenheimer — APPROVED** [review]

**Comment by marcoppenheimer** [comment]
Really neat and clean PR!

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 4 above)
Comment (zmraul) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:3054]: This fix: https://github.com/canonical/data-platform-libs/pull/163

**Pull request merged**